File: datastructures/linkedlist/java/algorithms/LinkedListCycleII.java
package datastructures.linkedlist.java.algorithms;

public class LinkedListCycleII {

    public static ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode intersect = findIntersect(head);
        if (intersect == null) {
            return null;
        }

        ListNode pointerHead = head;
        ListNode pointerIntersect = intersect;

        while (pointerHead != pointerIntersect) {
            pointerHead = pointerHead.next;
            pointerIntersect = pointerIntersect.next;
        }

        return pointerIntersect;
    }

    private static ListNode findIntersect(ListNode head) {
        ListNode slowPointer = head;
        ListNode fastPointer = head;

        while (fastPointer != null && fastPointer.next != null) {
            slowPointer = slowPointer.next;
            fastPointer = fastPointer.next.next;
            if (slowPointer == fastPointer) {
                return slowPointer;
            }
        }

        return null;
    }

    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);

        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node3;

        System.out.println(detectCycle(node1).val);
    }

}

File: datastructures/linkedlist/python/algorithms/linked-list-cycle-ii.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def get_intersect(head):
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

        if fast == slow:
            return slow

    return None


def detect_cycle(head):
    if not head:
        return None

    intersect = get_intersect(head)
    if not intersect:
        return None

    pointer_head = head
    pointer_intersect = intersect

    while pointer_head != pointer_intersect:
        pointer_head = pointer_head.next
        pointer_intersect = pointer_intersect.next

    return pointer_intersect


if __name__ == "__main__":
    node_1 = ListNode(1)
    node_2 = ListNode(2)
    node_3 = ListNode(3)
    node_4 = ListNode(4)

    node_1.next = node_2
    node_2.next = node_3
    node_3.next = node_4
    node_4.next = node_3

    print(detect_cycle(node_1).val)
